src/Uno.UI/DirectUI/PropertyPathStep.cs

```diff
@@ -0,0 +1,1546 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Data;
+using Uno.Disposables;
+using Windows.Foundation.Collections;
+
+namespace DirectUI;
+
+internal abstract partial class PropertyPathStep // src\dxaml\xcp\dxaml\lib\PropertyPathStep.h
+{
+	//public:
+	//public void Initialize(PropertyPathListener pListener);
+	//using ctl::WeakReferenceSource::Initialize;
+
+	//~PropertyPathStep() override;
+
+	public void SetNext(PropertyPathStep pNextStep)
+	{
+		m_tpNext = pNextStep;
+	}
+
+	public PropertyPathStep GetNextStep()
+	{
+		return m_tpNext;
+	}
+
+	public abstract void ReConnect(object pSource);
+
+	public abstract object GetValue();
+	public abstract void SetValue(object pValue);
+	public abstract Type GetSourceType();
+	public abstract bool IsConnected();
+
+	public abstract Type GetType2();
+
+	public virtual string DebugGetPropertyName() => null;
+
+	//protected:
+	//protected abstract void Disconnect();
+	protected virtual void DisconnectCurrentItem() { }
+	protected virtual void CollectionViewCurrentChanged() =>
+		// return E_NOTIMPL;
+		throw new NotImplementedException();
+
+	//protected void AddCurrentChangedEventHandler();
+
+	// This method is safe to be called from the destructor
+	//protected void SafeRemoveCurrentChangedEventHandler();
+
+	//protected void RaiseSourceChanged();
+
+	//protected:
+	protected PropertyPathStep m_tpNext;
+	protected WeakReference m_spListener;
+
+	//ctl::EventPtr<CurrentChangedEventCallback> m_epCurrentChangedHandler;
+	protected IDisposable m_epCurrentChangedHandler;
+	protected ICollectionView m_tpSourceAsCV;
+}
+partial class PropertyPathStep // src\dxaml\xcp\dxaml\lib\PropertyPathStep.cpp
+{
+	public void Initialize(PropertyPathListener pListener)
+	{
+		m_spListener = new WeakReference(pListener);
+	}
+
+	protected void Disconnect()
+	{
+		if (m_epCurrentChangedHandler is { })
+		{
+			m_epCurrentChangedHandler.Dispose();
+		}
+
+		m_tpSourceAsCV = null;
+
+		DisconnectCurrentItem();
+	}
+
+	~PropertyPathStep()
+	{
+		SafeRemoveCurrentChangedEventHandler();
+	}
```

Comment:
Let's try to avoid those finalizers, particularly as it may execute code on the UI thread